query/src/main/scala/filodb/query/exec/SetOperatorExec.scala

```diff
@@ -187,59 +188,175 @@ final case class SetOperatorExec(queryContext: QueryContext,
+          case (lhs, rhs) => (lhs, rvs)
+        }.getOrElse((Nil, rvs))
+        acc + (jk -> newVal)
+    }
+
+    case class OrSetOpIteratorWrapper(lhs: Seq[RangeVector], rhs: Seq[RangeVector]) {
+      // Has a side effect where LHS are consumed first and a bit is set to indicate the time instant where there
+      // is at least one non Nan value. the rhs values are then iterated over and emitted only where the given timestep
+      // has no value in any of the LHS timeseries
+      assert(outputRvRange.isDefined, "Expected to see a non null outputRvRange")
+      val (start, end, step) = (outputRvRange.get.startMs, outputRvRange.get.endMs, outputRvRange.get.stepMs)
+      // 64 bits in one word (long), 1 bit per step
+      assert(start == end || step > 0, "Expected to see a non 0 step size when start and end are not same")
+      val wordsRequired = if (start == end) 1 else ((((end - start) / step) >> 6) + 1).toInt
+      val bitSet = new mutable.BitSet(wordsRequired)
+
+      def joined(): Seq[RangeVector] = {
+        val lhsRvs: Seq[RangeVector] = lhs.map(x => {
+          var idx = 0
+          val rows = x.rows()
+          val rowsCursor: RangeVectorCursor = new RangeVectorCursor {
+            val cur = new TransientRow()
+
+            override def hasNext: Boolean = rows.hasNext
+
+            override def next(): RowReader = {
+              val reader = rows.next()
+              // TODO: What is the expected behavior for Histograms?
+              val value = reader.getDouble(1)
+              val res = if (value.isNaN) {
+                Double.NaN
+              } else {
+                bitSet += idx
+                value
+              }
+              idx += 1
+              cur.setValues(reader.getLong(0), res)
+              cur
+            }
+
+            override def close(): Unit = {
+              rows.close()
+            }
+          }
+          IteratorBackedRangeVector(x.key, rowsCursor, x.outputRange)
+        })
+
+        val rhsRvs: Seq[RangeVector] = rhs.map(x => {
+          var idx = 0
+          val rows = x.rows()
+          val rowsCursor: RangeVectorCursor = new RangeVectorCursor {
+            val cur = new TransientRow()
+
+            override def hasNext: Boolean = rows.hasNext
+
+            override def next(): RowReader = {
+              val reader = rows.next()
+              // TODO: What is the expected behavior for Histograms?
+              val res = if (bitSet.contains(idx)) {
+                Double.NaN
+              } else {
+                reader.getDouble(1)
+              }
+              idx += 1
+              cur.setValues(reader.getLong(0), res)
+              cur
+            }
+
+            override def close(): Unit = {
+              rows.close()
+            }
+          }
+          IteratorBackedRangeVector(x.key, rowsCursor, x.outputRange)
+        })
+        // This order is important, when lhs Rvs are executed they set a bitmap which is used when iterating RHS
```

Comment:
This part of code is same as the logic for the left. can we have a common function for it?

Reply:
Its not the same, the lhs is **setting** a bit and rhs is **checking** for the set bit and returning a NaN or a value. Am I missing something?